src/cart/crt_hg.c

```diff
@@ -589,6 +811,133 @@ crt_hg_init(crt_phy_addr_t *addr, bool server)
 	return rc;
 }
 
+int
+crt_dump_ha_list(d_list_t *halist)
+{
+	struct crt_ha_list_entry	*entry;
+	struct crt_ha_list_entry	*temp;
+	struct ha_entry			ha_value;
+	int				rc = 0;
+
+	d_list_for_each_entry_safe(entry, temp, halist,
+					chl_list_link) {
+		ha_value = entry->chl_entry;
+
+		D_DEBUG(DB_TRACE, "hg addr %" PRIu64 "\n",
+				(uint64_t)ha_value.ha_addr);
+		D_DEBUG(DB_TRACE, "hg class %" PRIu64 "\n\n",
+				(uint64_t) ha_value.ha_class);
+	}
+	return rc;
+}
+
+
+int
+crt_hg_ha_list_empty(d_list_t *halink, void *arg)
+{
+	struct crt_ha_mapping		*ha;
+	struct crt_ha_list_entry	*entry;
+	struct crt_ha_list_entry	*temp;
+	struct ha_entry			ha_value;
+	int				rc = 0;
+
+	D_ASSERT(arg == NULL);
+	ha = crt_ha_link2ptr(halink);
+
+	if (ha->chm_initialized != 1)
+		D_GOTO(out, rc);
+
+	D_DEBUG(DB_TRACE, "client id %" PRIu64 " removing from hash table\n",
+				ha->chm_key);
+
+	d_list_for_each_entry_safe(entry, temp, &ha->chm_list,
+					chl_list_link) {
+		ha_value = entry->chl_entry;
+		rc = crt_hg_remove_addr(ha_value.ha_class, ha_value.ha_addr);
+
+		if (rc != HG_SUCCESS) {
+			D_ERROR("crt_remove_addr failed, hg_ret %d.\n", rc);
+		}
+		D_FREE(entry);
+	}
+out:
+	return rc;
+}
+
+int
```

Comment:
it probably would help to add comment here to explain that this function is for server hash tables only